src/eda_ja/eda.py

```diff
@@ -1,204 +1,280 @@
-			'by', 'for', 'with', 'about', 'against', 'between',
-			'into', 'through', 'during', 'before', 'after', 
-			'above', 'below', 'to', 'from', 'up', 'down', 'in',
-			'out', 'on', 'off', 'over', 'under', 'again', 
-			'further', 'then', 'once', 'here', 'there', 'when', 
-			'where', 'why', 'how', 'all', 'any', 'both', 'each', 
-			'few', 'more', 'most', 'other', 'some', 'such', 'no', 
-			'nor', 'not', 'only', 'own', 'same', 'so', 'than', 'too', 
-			'very', 's', 't', 'can', 'will', 'just', 'don', 
-			'should', 'now', '']
-
-def get_only_chars(line):
-
-    clean_line = ""
-
-    line = line.replace("’", "")
-    line = line.replace("'", "")
-    line = line.replace("-", " ") #replace hyphens with spaces
-    line = line.replace("\t", " ")
-    line = line.replace("\n", " ")
-    line = line.lower()
-
-    for char in line:
-        if char in 'qwertyuiopasdfghjklzxcvbnm ':
-            clean_line += char
-        else:
-            clean_line += ' '
+import re
+import sqlite3
+from dataclasses import dataclass
 
-    clean_line = re.sub(' +',' ',clean_line) #delete extra spaces
-    if clean_line[0] == ' ':
-        clean_line = clean_line[1:]
-    return clean_line
+import MeCab
+import pandas as pd
+from nltk.corpus import wordnet
 
+from . import file_utils
 
 
-class EDA():
-    def __init__(self):
-        pass
+@dataclass
+class Words:
+    raw: list[str]
+    origin: list[str]
+
+    def __post_init__(self):
+        if not self.origin:
+            return
+        assert len(self.raw) == len(self.origin)
+
+    def __len__(self):
+        return len(self.raw)
+
+
+class Eda:
+    def __init__(self, stop_words_path):
+        self.stop_words = []
+        if self._has_stop_words(stop_words_path):
+            with open(stop_words_path) as f:
+                self.stop_words = [line.strip() for line in f.readlines()]
+
+    def _has_stop_words(self, stop_words_path):
+        if stop_words_path is None:
+            return False
+
+        if not os.path.isfile(stop_words_path):
+            return False
+
+        return True
+
+    def synonym_replacement(self, words: Words, n: int):
+        new_words = words.raw.copy()
+        random_word_indexes = [
+            i for i, word in enumerate(words.raw) if word not in self.stop_words
+        ]
+        random.shuffle(random_word_indexes)
 
-    def synonym_replacement(self, words, n):
-        new_words = words.copy()
-        random_word_list = list(set([word for word in words if word not in stop_words]))
-        random.shuffle(random_word_list)
         num_replaced = 0
-        for random_word in random_word_list:
-            synonyms = self.get_synonyms(random_word)
+        for index in random_word_indexes:
+            synonyms = self._get_synonyms(words.origin[index])
+
             if len(synonyms) >= 1:
                 synonym = random.choice(list(synonyms))
-                new_words = [synonym if word == random_word else word for word in new_words]
```

Comment:
all same words are replaced in original code, but i replace each words by index in order to reach replacement ratio to alpha_sr in config.